## Advent of Code

Advent of Code is an Advent calendar of small programming puzzles for a variety of skill sets and skill levels that can be solved in any programming language you like. People use them as a speed contest, interview prep, company training, university coursework, practice problems, or to challenge each other.

Visit: https://adventofcode.com/2021/about for more info

This workbook contains my work to solve the different problems from 2021

### Day 1 Pt 1

How many measurements are larger than the previous measurement?

In [1]:
import pandas as pd
import numpy as np
import random

from timeit import timeit
from itertools import repeat

In [2]:
df = pd.read_csv('AOC_list.csv')
df

Numbers  Inc/Dec
0         109      NaN
1         117      NaN
2         118      NaN
3          98      NaN
4         102      NaN
...       ...      ...
1995     6289      NaN
1996     6292      NaN
1997     6293      NaN
1998     6297      NaN
1999     6308      NaN

[2000 rows x 2 columns]

In [4]:
#goal: count the number of times a depth measurement increases
#from previous measurement

for i in df.index[1:2000]:
    if df['Numbers'][i-1] < df['Numbers'][i]:
        df['Inc/Dec'][i] = 'increasing'
    else:
        df['Inc/Dec'][i] = 'decreasing'
        
#returns count of increasing
df[df['Inc/Dec'] == 'increasing'].count()

C:\Users\jlucz\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\jlucz\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Numbers    1448
Inc/Dec    1448
dtype: int64

There are 1,448 instances of records increasing from the previous record.

### Day 1 Pt 2

Consider sums of a three-measurement sliding window. How many sums are larger than the previous sum?

In [5]:
#answer the same question, but based off a sliding window 
#of 3 measurements

df2 = pd.DataFrame(df['Numbers'])
df2['Sum'] = np.nan
df2['inc/dec'] = 'n/a'

#create window sum values and populated in df2

for i in df2.index[2:2000]:
    df2['Sum'][i] = (df2['Numbers'][i-2] + df2['Numbers'][i-1] +df2['Numbers'][i])
    
    #calc if window sum inc/dec
for i in df2.index[3:2000]:
    if df2['Sum'][i-1] < df2['Sum'][i]:
        df2['inc/dec'][i] = 'increasing'
    else:
        df2['inc/dec'][i] = 'decreasing'

df2[df2['inc/dec'] == 'increasing'].count()

C:\Users\jlucz\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\jlucz\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
C:\Users\jlucz\Anaconda3\lib\site-packages\ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  app.launch_new_instance()


Numbers    1471
Sum        1471
inc/dec    1471
dtype: int64

The resulting answer is 1,471 instances where the window sum of 3 numbers is increasing
